### PR TITLE
fix: handle multi namespace/section in foldingRange and documentSymbol

### PR DESCRIPTION
**Pull request by digama0**
In lean 4, unlike lean 3, `namespace`/`section` and `end` are not strictly paired -- the following nesting pattern is valid:
```lean
namespace O.A
namespace B
namespace C.D
end B.C.D
section E.F
end F
end A.E
```
This PR fixes the `documentSymbol` and `foldingRange` request handlers to handle this situation correctly, by breaking up start or end commands as necessary to create a well nested set of scopes, as if it were written:

```lean
namespace O namespace A
namespace B
namespace C.D
end C.D end B
section E section F
end F
end E end A

end O
```
For folding ranges, if you have multiple folding ranges starting on the same line only the first one is presented and the others are ignored in vscode, so we suppress all but the largest folding range starting at a given `namespace` or `section` command.

The `toDocumentSymbols` function is also refactored in this PR to be tail recursive and not make so many intermediate lists, making it resemble the `addRanges` function more closely.

### CODE REVIEW

File: src/Lean/Server/FileWorker/RequestHandling.lean [hunk 1]
```diff
@@ -440,16 +465,33 @@ partial def handleFoldingRange (_ : FoldingRangeParams)
     isImport stx := stx.isOfKind ``Lean.Parser.Module.header || stx.isOfKind ``Lean.Parser.Command.open
 
     addRanges (text : FileMap) sections
-    | [] => return
+    | [] => do
+      if let (_, start)::rest := sections then
+        addRange text FoldingRangeKind.region start text.source.endPos
+        addRanges text rest []
+      else
+        return
```
Comment (Kha) [inline src/Lean/Server/FileWorker/RequestHandling.lean:473]: Can be elided, no?

File: src/Lean/Server/FileWorker/RequestHandling.lean [hunk 2]
```diff
@@ -272,20 +295,38 @@ partial def handleDocumentSymbol (_ : DocumentSymbolParams)
...
+      Array DocumentSymbol :=
     match stxs with
-    | [] => ([], [])
+    | [] => stack.foldl (fun syms entry => entry.finish text syms none) syms
     | stx::stxs => match stx with
-      | `(namespace $id)  => sectionLikeToDocumentSymbols text stx stxs (id.getId.toString) SymbolKind.namespace id
-      | `(section $(id)?) => sectionLikeToDocumentSymbols text stx stxs ((·.getId.toString) <$> id |>.getD "<section>") SymbolKind.namespace (id.map (·.raw) |>.getD stx)
-      | `(end $(_id)?) => ([], stx::stxs)
+      | `(namespace $id)  =>
+        let entry := { name := id.getId.components', stx, selection := id, prevSiblings := syms }
```
Comment (Kha) [inline src/Lean/Server/FileWorker/RequestHandling.lean:307]: OMG, we should definitely rename `components'` to `componentsRev`. Your `foldr` was a complete mystery to me before I realized that, there should be a comment on *why* we store `name` like that.

File: src/Lean/Server/FileWorker/RequestHandling.lean [hunk 3]
```diff
@@ -272,20 +295,38 @@ partial def handleDocumentSymbol (_ : DocumentSymbolParams)
...
+        toDocumentSymbols text stxs #[] (entry :: stack)
+      | `(section $(id)?) =>
+        let name := id.map (·.getId.components') |>.getD [`«»]
```
Comment (Kha) [inline src/Lean/Server/FileWorker/RequestHandling.lean:310]: This is to differentiate it from `.anonymous` from a partial syntax tree I assume?

**Review by Kha — APPROVED** [review]
LGTM, but more comments and docstrings for the tricky parts would be appreciated!

File: src/Lean/Server/FileWorker/RequestHandling.lean (hunk 3 above)
Comment (digama0) [reply src/Lean/Server/FileWorker/RequestHandling.lean:310]: This isn't an arbitrary choice: anonymous `section` is literally equivalent to `section «»`, and 
```lean
section A
section 
namespace B
end A.«».B
```
is legal, cursed though it is.

**Pull request merged**